src/api.ts

```diff
@@ -0,0 +1,187 @@
+import fs from 'fs';
+import path from 'path';
+import glob from 'glob';
+
+import { generateHtml } from './html';
+import { exploreBundle, UNMAPPED_KEY } from './explore';
+import { AppError, getErrorMessage } from './app-error';
+import {
+  BundlesAndFileTokens,
+  ExploreOptions,
+  ExploreResult,
+  Bundle,
+  ExploreErrorResult,
+  ExploreBundleResult,
+  ErrorCode,
+} from './index';
+
+/**
+ * Analyze bundle(s)
+ */
+export async function explore(
+  bundlesAndFileTokens: BundlesAndFileTokens,
+  options: ExploreOptions = {}
+): Promise<ExploreResult> {
+  const bundles: Bundle[] = [];
+  const fileTokens: string[] = [];
+
+  bundlesAndFileTokens = Array.isArray(bundlesAndFileTokens)
+    ? bundlesAndFileTokens
+    : [bundlesAndFileTokens];
+
+  if (bundlesAndFileTokens.length === 0) {
+    throw new AppError(ErrorCode.NoBundles);
+  }
+
+  // Separate bundles from file tokens
+  bundlesAndFileTokens.forEach(item => {
+    if (typeof item === 'string') {
+      fileTokens.push(item);
+    } else {
+      bundles.push(item);
+    }
+  });
+
+  // Get bundles from file tokens
+  bundles.push(...getBundles(fileTokens));
+
+  return Promise.all(
+    bundles.map(bundle =>
+      exploreBundle(bundle, options).catch<ExploreErrorResult>(error =>
+        onExploreError(bundle, error)
+      )
+    )
+  )
+    .then(results => getExploreResult(results, options))
+    .then(result => {
+      // Reject if none of results is successful
+      if (result.bundles.length === 0) {
+        return Promise.reject(result);
+      }
+
+      saveHtmlToFile(result, options);
+
+      return result;
+    });
+}
+
+/**
+ * Expand list of file tokens into a list of bundles
+ */
+export function getBundles(fileTokens: string[]): Bundle[] {
+  const filenames = fileTokens.reduce<string[]>((result, filePath) => {
+    if (glob.hasMagic(filePath)) {
+      result.push(...expandGlob(filePath));
+    } else {
+      result.push(filePath);
+    }
+
+    return result;
+  }, []);
+
+  const codeFilenames: string[] = [];
+  const mapFilenames: string[] = [];
+
+  filenames.forEach(filename => {
```

Comment:
This pattern appears a few times. Maybe factor out a helper like

```ts
function separateBy<T>(xs: T[], pred: (x: T) => boolean): [T[], T[]] {
  // ...
}
```

and then

```ts
const [codeFilenames, mapFilenames] = separateBy(filenames, f => f.endsWith('.map'));
```

Reply:
Agree

Reply:
I will use `_.partition`

Reply:
Done